core/server/services/members/importer/index.js

```diff
@@ -4,50 +4,153 @@ const ObjectId = require('bson-objectid');
-    if (_.isString(labels)) {
-        if (labels === '') {
-            return [];
+const CHUNK_SIZE = 5000;
+
+async function insert(table, data) {
+    const result = {
+        successful: 0,
+        unsuccessful: 0,
+        errors: []
+    };
+    for (const chunk of _.chunk(data, CHUNK_SIZE)) {
+        try {
+            await db.knex(table).insert(chunk);
+            result.successful += chunk.length;
+        } catch (error) {
+            result.unsuccessful += chunk.length;
+            result.errors.push(error);
         }
+    }
+    return result;
+}
 
-        return [{
-            name: labels.trim()
-        }];
-    } else if (labels) {
-        return labels.filter((label) => {
-            return !!label;
-        }).map((label) => {
-            if (_.isString(label)) {
-                return {
-                    name: label.trim()
-                };
-            }
-            return label;
-        });
+async function del(table, ids) {
+    const result = {
+        successful: 0,
+        unsuccessful: 0,
+        errors: []
+    };
+    for (const chunk of _.chunk(ids, CHUNK_SIZE)) {
+        try {
+            await db.knex(table).whereIn('id', chunk).del();
+            result.successful += chunk.length;
+        } catch (error) {
+            result.unsuccessful += chunk.length;
+            result.errors.push(error);
+        }
     }
-    return [];
+    return result;
 }
 
-const doImport = async ({members, allLabelModels, importSetLabels, imported, invalid, createdBy}) => {
-    const mappedMemberBatchData = [];
-    const mappedMembersLabelsBatchAssociations = [];
-    const membersWithStripeCustomers = [];
-    const membersWithComplimentaryPlans = [];
+function serializeMemberLabels(labels) {
+    return labels.reduce((labelsAcc, label) => {
+        if (!label) {
+            return labels;
+        }
+        if (typeof label !== 'string') {
+            return labelsAcc.concat(label.name);
+        }
+        return labelsAcc.concat(label);
+    });
+}
 
-    members.forEach((member) => {
-        cleanupUndefined(member);
+function getMemberData({members, allLabelModels, importSetLabels, createdBy}) {
+    const labelIdLookup = allLabelModels.reduce(function (labelIdLookupAcc, labelModel) {
+        return Object.assign(labelIdLookupAcc, {
+            [labelModel.get('name')]: labelModel.id
+        });
+    }, {});
+
+    const importedLabels = importSetLabels.map(label => label.name);
+
+    const stripeIsConnected = membersService.config.isStripeConnected();
+
+    const invalidMembers = [];
+    const membersToInsert = [];
+    const stripeCustomersToInsert = [];
+    const labelAssociationsToInsert = [];
+
+    members.forEach(function (member) {
+        if (member.stripe_customer_id && !stripeIsConnected) {
+            invalidMembers.push(member);
+            return;
+        }
+
+        // @TODO This is expensive, maybe we can just error if we get shoddy data?
```

Comment:
I'd even be up for dropping this whole block because it was added to handle a bug that happened in members export [pre version 3.6](https://github.com/TryGhost/Ghost/blob/78911257f407da311cf6a62bc148a51e5de4a4fb/core/server/api/canary/members.js#L16-L19) - limited time window. This is just my opinion, would need verification!

Reply:
Okay, maybe one for @matthanley to take a look at?

TL;DR pre 3.6 we would export the string 'undefined' for undefined values, and we now have to loop through every key in every object to handle that - which is expensive, especially for large imports.

Would you consider dropping the support for that?  We could provide a command/tool to clean the data up locally if necessary OR we could offload that work to Ghost-Admin, that way the operation would not affect the request cycle